### PR TITLE
Handle `mask` in `TimeDistributed` wrapper.

### PR DESCRIPTION
**Pull request by PeterChe1990**
This PR aims to add proper support of `mask` in `TimeDistributed` wrapper layer. Several previous PRs have raised questions about this or have attempted to solve it, but I didn't find any perfect solution so far. I may not be able to have a perfect one either but am willing to improve it. The most frequent usage of such layers with masking support might be a hierarchical RNN on a sentence with words with characters, or a paragraph with sentences with words. (E.g., models like [this](https://github.com/keras-team/keras/blob/master/examples/mnist_hierarchical_rnn.py)).

Here are the main ideas/purposes of this pr:

1) For layers like `TimeDistributed(Embedding(..., mask_zero=True))` or `TimeDistributed(Masking(...))`, the layer should call `compute_mask()` at each time step and stack them.
2) For layers like `TimeDistributed(LSTM(...))`, the `input_mask` should be split and fed into the `LSTM` layer at each time step. And if the `output_shape` changes (e.g., `return_sequences=False` in `LSTM`), the `output_mask` of the `TimeDistributed` layer should also change correspondingly.
3) The `TimeDistributed` layer should support, e.g., `input_shape=(None, None)`. For example, the number of words in a sentence and the number of characters in a word can be different from sample to sample. Currently the `TimeDistributed` can only support one unspecified input dimension.

Here are some details of this PR to discuss, which might have better solutions:
1) `K.int_shape()` may not be available for some `mask` with `Theano` backend, especially when the inner layer is `Embedding` or `Masking`. This is because `_keras_shape` is not set in lots of operations in `theano_backend` if the tensor shape is not changed. I manually create the `_keras_shape` in `compute_mask` of these two layers.
2) When `input_shape[0] is not None` in `TimeDistributed` layer, we need to use `K.rnn()` to dynamically call the inner layer. Because `K.rnn()` can only take 1 tensor input but we need to feed both `input` and `mask`, I didn't do anything in this situation. Probably we can pass all `mask` as constants to `K.rnn()` but I was expecting better solutions.
3) If there's `None` from the output of `K.int_shape`, we need to replace it by the output from `K.shape`, so that `reshape` can be correctly called.

### CODE REVIEW

File: tests/keras/layers/wrappers_test.py [hunk 1]
```diff
@@ -156,6 +156,59 @@ def test_TimeDistributed_trainable():
...
+    ref_mask_val_1 = ref_mask_val_0          # first RNN layer
+    ref_mask_val_2 = np.any(ref_mask_val_1, axis=-1)     # second RNN layer
+    ref_mask_val = [ref_mask_val_0, ref_mask_val_1, ref_mask_val_2]
+    for i in range(3):
+        assert np.array_equal(mask_outputs_val[i], ref_mask_val[i])
+    assert mask_outputs[-1] is None  # final layer
+
+
+@keras_test
+def test_TimeDistributed_with_mask_and_unspecified_shape():
```
Comment (fchollet) [inline tests/keras/layers/wrappers_test.py]: The name of this test is the same as the previous one (copy and paste issue?)

File: keras/layers/core.py [hunk 2]
```diff
@@ -60,7 +60,10 @@ def __init__(self, mask_value=0., **kwargs):
         self.mask_value = mask_value
 
     def compute_mask(self, inputs, mask=None):
-        return K.any(K.not_equal(inputs, self.mask_value), axis=-1)
+        output_mask = K.any(K.not_equal(inputs, self.mask_value), axis=-1)
+        if hasattr(inputs, '_keras_shape'):
+            output_mask._keras_shape = inputs._keras_shape[:-1]
+        return output_mask
```
Comment (fchollet) [inline keras/layers/core.py]: This change is specific to static shape inference in Theano. It should be made in `K.any` and `K.not_equal` (only in the Theano backend).

File: keras/layers/embeddings.py [hunk 3]
```diff
@@ -108,8 +109,10 @@ def build(self, input_shape):
     def compute_mask(self, inputs, mask=None):
         if not self.mask_zero:
             return None
-        else:
-            return K.not_equal(inputs, 0)
+        output_mask = K.not_equal(inputs, 0)
+        if hasattr(inputs, '_keras_shape'):
+            output_mask._keras_shape = inputs._keras_shape
+        return output_mask
```
Comment (fchollet) [inline keras/layers/embeddings.py]: Likewise, this change should be made in the Theano backend (`not_equal`).

File: keras/layers/wrappers.py [hunk 4]
```diff
@@ -227,6 +246,59 @@ def step(x, _):
...
+            for _ in range(2, len(K.int_shape(mask))):
+                output_mask = K.any(output_mask, axis=-1)
+        else:
+            # output_mask is not None. We need to reshape it
+            input_length = input_shape[1]
+            if not input_length:
+                input_length = K.shape(inputs)[1]
+            output_mask_int_shape = K.int_shape(output_mask)
+            if output_mask_int_shape is None:
+                # if the output_mask does not have `_keras_shape`,
```
Comment (fchollet) [inline keras/layers/wrappers.py]: No reference to `_keras_shape`, that's an implementation detail. Call it "static shape"

File: keras/layers/wrappers.py [hunk 5]
```diff
@@ -227,6 +246,59 @@ def step(x, _):
             y._uses_learning_phase = True
         return y
 
+    def compute_mask(self, inputs, mask=None):
```
Comment (fchollet) [inline keras/layers/wrappers.py]: Please add a docstring describing the overall strategy used to compute the mask.

File: keras/layers/wrappers.py [hunk 6]
```diff
@@ -201,21 +201,40 @@ def step(x, _):
             # No batch size specified, therefore the layer will be able
             # to process batches of any size.
             # We can go with reshape-based implementation for performance.
+
+            def get_shape_tuple(tensor, start_idx, init_tuple, int_shape=None):
```
Comment (fchollet) [inline keras/layers/wrappers.py]: Please add a docstring explaining the purpose of this method.

**Comment by PeterChe1990** [comment]
Thanks for the comments! I made some corresponding changes in the new commit.

**Pull request merged**